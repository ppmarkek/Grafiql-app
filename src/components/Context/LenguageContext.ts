import { Dispatch, SetStateAction, createContext } from 'react';

export enum Langs {
  en = 'en',
  ru = 'ru',
}

interface LenguageContextType {
  language: Langs;
  setLanguage: Dispatch<SetStateAction<Langs>>;
}

export const initialState: LenguageContextType = {
  language: Langs.en,
  setLanguage: (() => {}) as Dispatch<SetStateAction<Langs>>,
};

export const LanguageContext = createContext<LenguageContextType>(initialState);
