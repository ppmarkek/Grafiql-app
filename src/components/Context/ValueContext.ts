import { Dispatch, SetStateAction, createContext, useContext } from 'react';

import en from '../../i18n/en.json';
import ru from '../../i18n/ru.json';

export enum Langs {
  en = 'en',
  ru = 'ru',
}

interface ValueContextType {
  language: Langs;
  setLanguage: Dispatch<SetStateAction<Langs>>;
  inputEntryPoint: string;
  setInputEntryPoint: Dispatch<SetStateAction<string>>;
}

export const initialState: ValueContextType = {
  language: Langs.ru,
  setLanguage: (() => {}) as Dispatch<SetStateAction<Langs>>,
  inputEntryPoint: '',
  setInputEntryPoint: (() => {}) as Dispatch<SetStateAction<string>>,
};

export const ValueContext = createContext<ValueContextType>(initialState);

export function useLanguage(): Langs {
  const { language } = useContext<ValueContextType>(ValueContext);

  return language;
}

export function useI18n() {
  const { language } = useContext<ValueContextType>(ValueContext);
  switch (language) {
    case 'ru':
      return ru;
    default:
    case 'en':
      return en;
  }
}
