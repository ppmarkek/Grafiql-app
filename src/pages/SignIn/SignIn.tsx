import { Button, Grid, List, ListItem, Paper } from '@mui/material';
import { FormikProps, useFormik } from 'formik';
import FormInput, { Values } from '../../components/FormInput/FormInput';
import i18n from '../SignUp/en.json';

const SignIn = () => {
  const formik: FormikProps<Values> = useFormik({
    initialValues: {
      name: '',
      email: '',
      password: '',
      passwordConfirm: '',
    },
    onSubmit: (values) => {
      alert(JSON.stringify(values, null, 2));
    },
  });
  return (
    <Grid
      sx={{
        width: '100%',
        height: '100%',
        position: 'relative',
        display: 'flex',
        alignItems: 'center',
      }}
    >
      <Paper
        elevation={3}
        sx={{
          width: '60vw',
          height: '60vh',
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-around',
          alignItems: 'center',
          position: 'relative',
          margin: 'auto',
        }}
      >
        <h1>{i18n.auth.signIn}</h1>
        <form onSubmit={formik.handleSubmit}>
          <List
            sx={{ width: '100%', maxWidth: 360, bgcolor: 'background.paper' }}
          >
            <ListItem>
              <FormInput
                title={'email'}
                text={i18n.auth.email}
                type={'email'}
                formik={formik}
              />
            </ListItem>
            <ListItem>
              <FormInput
                title={'password'}
                text={i18n.auth.password}
                type={'password'}
                formik={formik}
              />
            </ListItem>
            <ListItem>
              <Button type="submit">{i18n.auth.buttonName}</Button>
            </ListItem>
          </List>
        </form>
      </Paper>
    </Grid>
  );
};

export default SignIn;
