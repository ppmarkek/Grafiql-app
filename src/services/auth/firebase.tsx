import { FirebaseApp, FirebaseError, initializeApp } from 'firebase/app';
import {
  GoogleAuthProvider,
  getAuth,
  signInWithPopup,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  User,
} from 'firebase/auth';
import {
  getFirestore,
  query,
  getDocs,
  collection,
  where,
  addDoc,
} from 'firebase/firestore';
import {
  Dispatch,
  ReactNode,
  createContext,
  useEffect,
  useReducer,
  useContext,
} from 'react';

interface FirebaseAuthState {
  user?: User;
  loading: boolean;
  error?: FirebaseError;
  // firstCheck: boolean
}
interface FirebaseContext {
  user?: User;
  loading: boolean;
  error?: FirebaseError;
  firebase: FirebaseApp;
  setState: Dispatch<FirebaseAuthState>;
}

const FirebaseContext = createContext<FirebaseContext | null>(null);

const firebaseConfig = {
  apiKey: process.env.REACT_APP_API_KEY,
  appId: process.env.REACT_APP_APP_ID,
  authDomain: process.env.REACT_APP_AUTH_DOMAIN,
  measurementId: process.env.REACT_APP_MEASUREMENT_ID,
  projectId: process.env.REACT_APP_PROJECT_ID,
  storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
};
const appFirebase = initializeApp(firebaseConfig);

export function FirebaseAuthProvider({
  children,
}: {
  children: ReactNode;
}): JSX.Element {
  const auth = getAuth(appFirebase);
  const firebaseCurrentUser = auth.currentUser;
  const useCustomUseReducer = (initState: FirebaseAuthState) => {
    const newState = useReducer(
      (prev: FirebaseAuthState, curr: FirebaseAuthState) => {
        return { ...prev, ...curr };
      },
      initState
    );
    return newState;
  };
  const [{ user, loading, error }, setState] = useCustomUseReducer({
    user: firebaseCurrentUser === null ? undefined : firebaseCurrentUser,
    loading: false,
    firstCheck: false,
  } as FirebaseAuthState);

  useEffect(() => {
    !user && setState({ loading: true });
    return auth.onAuthStateChanged(function (user: User | null) {
      setState({
        loading: false,
        user: user === null ? undefined : user,
      });
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth, setState]);

  return (
    <FirebaseContext.Provider
      value={{
        user,
        loading,
        error,
        firebase: appFirebase,
        setState,
      }}
    >
      {children}
    </FirebaseContext.Provider>
  );
}

export function useFirebaseAuth() {
  const firebaseContext = useContext(FirebaseContext);

  if (firebaseContext === null) {
    throw new Error('No FirebaseAuthProvider found.');
  }

  const { user, loading, error, setState, firebase } = firebaseContext;

  const logInWithEmailAndPassword = async (email: string, password: string) => {
    try {
      setState({ loading: true });
      const auth = getAuth(firebase);
      await signInWithEmailAndPassword(auth, email, password);
      setState({ loading: false });
    } catch (err) {
      console.error(err);
      if (err instanceof FirebaseError) {
        setState({
          error: err,
          loading: false,
        });
      }
    }
  };

  const logout = async () => {
    setState({ loading: true });
    const auth = getAuth(firebase);
    setState({ loading: false });
    return signOut(auth);
  };

  const registerWithEmailAndPassword = async (
    name: string,
    email: string,
    password: string
  ) => {
    try {
      const auth = getAuth(firebase);
      setState({ loading: true });
      const res = await createUserWithEmailAndPassword(auth, email, password);
      const user = res.user;
      const db = getFirestore(firebase);
      await addDoc(collection(db, 'users'), {
        uid: user.uid,
        name,
        authProvider: 'local',
        email,
      });
      setState({ loading: false });
    } catch (err) {
      if (err instanceof FirebaseError) {
        setState({
          error: err,
          loading: false,
        });
      }
    }
  };

  const signInWithGoogle = async () => {
    const googleProvider = new GoogleAuthProvider();
    try {
      const auth = getAuth(firebase);
      setState({ loading: true });
      const res = await signInWithPopup(auth, googleProvider);
      const user = res.user;
      const db = getFirestore(firebase);
      const q = query(collection(db, 'users'), where('uid', '==', user.uid));
      const docs = await getDocs(q);
      if (docs.docs.length === 0) {
        await addDoc(collection(db, 'users'), {
          uid: user.uid,
          name: user.displayName,
          authProvider: 'google',
          email: user.email,
        });
      }
      setState({ loading: false });
    } catch (err) {
      if (err instanceof FirebaseError) {
        setState({
          error: err,
          loading: false,
        });
      }
    }
  };

  return {
    user,
    loading,
    error,
    registerWithEmailAndPassword,
    logInWithEmailAndPassword,
    logout,
    signInWithGoogle,
  };
}
