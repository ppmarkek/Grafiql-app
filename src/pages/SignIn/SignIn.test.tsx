import React from 'react';
import SignIn from './SignIn';
import renderer from 'react-test-renderer';
import * as FirebaseModule from '../../services/auth/firebase';
import * as useI18nModule from '../../components/Context/ValueContext';
import lang from '../../i18n/en.json';

type useFirebaseAuthREturn = ReturnType<typeof FirebaseModule.useFirebaseAuth>;
const mockedReturnedValue: useFirebaseAuthREturn = {
  user: undefined,
  error: undefined,
  loading: false,
  registerWithEmailAndPassword: jest.fn(),
  logInWithEmailAndPassword: jest.fn(),
  logout: jest.fn(),
  signInWithGoogle: jest.fn(),
};
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
  Link: () => <a></a>,
}));

describe('SignIn', () => {
  beforeEach(() => {
    jest
      .spyOn(FirebaseModule, 'useFirebaseAuth')
      .mockReturnValue(mockedReturnedValue);
    jest.spyOn(useI18nModule, 'useI18n').mockReturnValue(lang);
  });
  it('should render without errors', () => {
    const tree = renderer.create(<SignIn />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
