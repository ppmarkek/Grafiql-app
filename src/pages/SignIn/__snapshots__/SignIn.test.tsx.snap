// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn should render without errors 1`] = `
<div
  className="MuiGrid-root css-x9vj63-MuiGrid-root"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-dfu1h7-MuiPaper-root"
  >
    <h1>
      SignIn
    </h1>
    <form
      onSubmit={[Function]}
    >
      <ul
        className="MuiList-root MuiList-padding css-1mcnwpj-MuiList-root"
      >
        <li
          className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
          disabled={false}
        >
          <div
            className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="email"
            >
              email
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-describedby="component-error-text"
                aria-invalid={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="email"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="email"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    email
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
          disabled={false}
        >
          <div
            className="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink={false}
              htmlFor="password"
            >
              password
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              onClick={[Function]}
            >
              <input
                aria-describedby="component-error-text"
                aria-invalid={false}
                className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled={false}
                id="password"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="password"
                value=""
              />
              <fieldset
                aria-hidden={true}
                className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="css-yjsfm1"
                >
                  <span>
                    password
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </li>
        <li
          className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
          disabled={false}
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            onBlur={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            Submit
          </button>
        </li>
      </ul>
    </form>
  </div>
</div>
`;
